src/renderer.c

```diff
@@ -120,84 +164,349 @@ static int font_set_style(FT_Outline* outline, int x_translation, unsigned char
+    int flags = GLYPH_LOADED_NORMAL;
+    float whitespace_advance = font_get_whitespace_advance(font, codepoint);
+
+    if (whitespace_advance < 0) {
+      // not whitespace
+      flags = GLYPH_LOADED_MISSING;
+      whitespace_advance = 0;
+    }
+
+    for (int j = 0; j < bitmaps_cached; ++j) {
+      font->sets[j][glyphset_idx]->metrics[metrics_idx].flags = flags;
+      font->sets[j][glyphset_idx]->metrics[metrics_idx].xadvance = whitespace_advance;
+    }
+
+    if (flags == GLYPH_LOADED_NORMAL || font->missing_glyph_codepoint != codepoint)
+      return;
+  }
+
+  /**
+   * There's an issue with monospaced fonts where the hinted xadvance is off (rounded)
+   * when calculated with lsb_delta and rsb_delta;
+   * we need to load the font without hinting and use that.
+   * See #843.
+   */
+  if (FT_Load_Glyph(font->face, glyph_index, (load_flags | FT_LOAD_BITMAP_METRICS_ONLY | FT_LOAD_NO_HINTING) & ~FT_LOAD_FORCE_AUTOHINT))
+    return;
+
+  // subpixel positioning & transformations don't affect xadvance, so we don't run them
+  for (int j = 0; j < bitmaps_cached; ++j) {
+    // since we already have the xadvance, we can consider this codepoint loaded
+    font->sets[j][glyphset_idx]->metrics[metrics_idx].flags = GLYPH_LOADED_NORMAL;
+    font->sets[j][glyphset_idx]->metrics[metrics_idx].xadvance = font->face->glyph->advance.x / 64.0f;
+  }
+
+  for (int j = 0; j < bitmaps_cached; ++j) {
+
+    int width, height;
+    FT_GlyphSlot slot;
+    uint8_t *src_pixels, *dst_pixels;
+    GlyphSet *set = font->sets[j][glyphset_idx];
+    GlyphMetric *metric = &set->metrics[metrics_idx];
+
+    // we must reload the glyph instead of incrementally translating it because FT_Render_Glyph
+    // will not re-render the glyph unless the slot is "cleared".
+    if (FT_Load_Glyph(font->face, glyph_index, load_flags))
+      return;
+
+    slot = font->face->glyph;
+    if (slot->format == FT_GLYPH_FORMAT_OUTLINE) {
+      if (slot->outline.n_points == 0) continue;
+      // perform style and subpixel positioning
+      font_set_style(&slot->outline, font->style);
+      if (j > 0)
+        FT_Outline_Translate(&slot->outline, j * (64 / SUBPIXEL_BITMAPS_CACHED), 0);
+    }
+
+    if (FT_Render_Glyph(slot, render_flags))
+      return;
+
+    width = slot->bitmap.width;
+    height = slot->bitmap.rows;
+    if (slot->bitmap.pixel_mode == FT_PIXEL_MODE_LCD)
+      width /= 3;
+
+    if (width == 0 || height == 0)
+      return;
+
+    // save bitmap metrics
+    metric->x0 = 0;
+    metric->x1 = width;
+    metric->y0 = set->pen_y;
+    metric->y1 = metric->y0 + height;
+    metric->bitmap_left = slot->bitmap_left;
+    metric->bitmap_top = slot->bitmap_top;
+
+    if (!set->surface || metric->x1 > set->surface->w || metric->y1 > set->surface->h) {
+      // resize the glyphset with exact pixel sizing
+      font_allocate_glyphset(font, glyphset_idx, false);
+      assert(metric->x1 <= set->surface->w && metric->y1 <= set->surface->h);
```

Comment:
Just in case, the assert will cause a segfault if `!set->surface`  since it is trying to access the uninitialized field `surface`.  This escaped my eyes when applying the previous workaround which is corrected now in pragtical/pragtical#114